### PR TITLE
The `-test` command for windows. Nuget packages

### PR DESCRIPTION
**Pull request by michaelgsharp**
The `-test` command works for windows. CI doesn't work yet as the CI pipeline hasn't been updated. Neither does nightly.

Nuget packages can now be created. Symbols packages are not working yet.

### CODE REVIEW

File: Directory.Build.targets [hunk 1]
```diff
@@ -1,4 +1,46 @@
...
+			DestinationFolder="$(OutDir)"
+			OverwriteReadOnlyFiles="$(OverwriteReadOnlyFiles)"
+			Retries="$(CopyRetryCount)"
+			RetryDelayMilliseconds="$(CopyRetryDelayMilliseconds)"
+			UseHardlinksIfPossible="$(CreateHardLinksForPublishFilesIfPossible)"
+			UseSymboliclinksIfPossible="$(CreateSymbolicLinksForPublishFilesIfPossible)">
+		<Output TaskParameter="DestinationFiles" ItemName="FileWrites"/>
+		</Copy>
+  <!--<Exec Command="call C:\MyFile.bat or call C:\MyFile.bat call C:\MyFile2.bat" />-->
+  </Target>
```
Comment (harishsk) [inline Directory.Build.targets:42]: nit: Please delete this

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj [hunk 2]
```diff
@@ -7,6 +7,14 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\..\src\Microsoft.ML.AutoML\Microsoft.ML.AutoML.csproj" />
+    <ProjectReference Include="..\..\..\src\Microsoft.ML.Core\Microsoft.ML.Core.csproj" >
+      <PrivateAssets>all</PrivateAssets>
```
Comment (harishsk) [inline docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:11]: question: Why is this necessary with Arcade?

File: docs/samples/Microsoft.ML.Samples.GPU/Microsoft.ML.Samples.GPU.csproj [hunk 3]
```diff
@@ -20,9 +20,13 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\..\src\Microsoft.ML.Vision\Microsoft.ML.Vision.csproj" />
+    <ProjectReference Include="..\..\..\src\Microsoft.ML.ImageAnalytics\Microsoft.ML.ImageAnalytics.csproj" />
+    <ProjectReference Include="..\..\..\src\Microsoft.ML.Data\Microsoft.ML.Data.csproj" />
```
Comment (harishsk) [inline docs/samples/Microsoft.ML.Samples.GPU/Microsoft.ML.Samples.GPU.csproj:24]: Interesting...it seems like we are requiring more project references than before. Any idea why?

File: eng/Build.props [hunk 4]
```diff
@@ -1,8 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project>
+
+  <ItemDefinitionGroup>
+    <ProjectToBuild>
+      <RestoreInParallel>true</RestoreInParallel>
+      <BuildInParallel>false</BuildInParallel>
+    </ProjectToBuild>
```
Comment (harishsk) [inline eng/Build.props:8]: question: Where does it fail if we enable parallel builds?

File: pkg/Directory.Build.props [hunk 5]
```diff
@@ -1,21 +1,14 @@
...
-  <Import Project="..\Directory.Build.props" />
 
   <PropertyGroup>
-    <NoBuild>true</NoBuild>
-    <IncludeBuildOutput>false</IncludeBuildOutput>
-    <IncludeSymbols>false</IncludeSymbols>
-
-    <IsSymbolsPackage Condition="$(MSBuildProjectName.Contains('.symbols'))">true</IsSymbolsPackage>
+    <PackageAssetsPath>$(ArtifactsDir)pkgassets/</PackageAssetsPath>
+    <IncludeSymbols>true</IncludeSymbols>
```
Comment (harishsk) [inline pkg/Directory.Build.props:5]: Are we still using the pkg folder and the nupkg projects? I thought they were going to be obsoleted.

File: src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj [hunk 6]
```diff
@@ -1,19 +1,29 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="$(RepoRoot)pkg/Directory.Build.props" />
+
```
Comment (harishsk) [inline src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj:3]: Are we using the pkg directory only for the Directory.Build.props to set package properties? 
Are we using anything else from the package directory? If not, would it make sense to move that props file out of there and delete the pkg directory?

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs [hunk 7]
```diff
@@ -1110,11 +1110,11 @@ private void LoadExtensions(Assembly assembly, bool throwOnError)
         /// </remarks>
```
Comment (harishsk) [inline src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: I am not entirely convinced this is a valid test. I don't see any reason why some of the ML.NET product assemblies can't contain extensions. The HasMLNetPublicKey test is also no longer valid because product and shipping assemblies are now signed with the same key. 
Can we delete the related code?

File: test/Directory.Build.targets [hunk 8]
```diff
@@ -3,9 +3,10 @@
...
 
   <PropertyGroup>
-    <!-- This needs to be in Directory.Build.targets or the default value for AllowedReferenceRelatedFileExtensions will not be applied. 
+    <!-- This needs to be in Directory.Build.targets or the default value for AllowedReferenceRelatedFileExtensions will not be applied.
          This is required in order to have following extension files in the output folder of the projects referencing other projects. -->
-    <AllowedReferenceRelatedFileExtensions>$(AllowedReferenceRelatedFileExtensions);.runtimeconfig.json;.runtimeconfig.dev.json;.deps.json</AllowedReferenceRelatedFileExtensions>  
+    <AllowedReferenceRelatedFileExtensions>$(AllowedReferenceRelatedFileExtensions);.runtimeconfig.json;.runtimeconfig.dev.json;.deps.json</AllowedReferenceRelatedFileExtensions>
   </PropertyGroup>
 
+
```
Comment (harishsk) [inline test/Directory.Build.targets:11]: nit: remove blank line

File: tools-local/Microsoft.ML.StableApi/Microsoft.ML.StableApi.csproj [hunk 9]
```diff
@@ -6,23 +6,23 @@
...
-    <PackageReference Include="Microsoft.ML.FastTree" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.LightGbm" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.ImageAnalytics" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.Mkl.Components" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.TimeSeries" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.TensorFlow" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.OnnxTransformer" Version="1.4.0" />
-    <PackageReference Include="Microsoft.ML.Vision" Version="1.4.0" />
+    <PackageReference Include="Microsoft.Extensions.ML" Version="1.5.2" />
+    <PackageReference Include="Microsoft.ML" Version="1.5.2" />
```
Comment (harishsk) [inline tools-local/Microsoft.ML.StableApi/Microsoft.ML.StableApi.csproj:10]: We should probably set this to 1.5.0? If we consider the final .x to be a patch release, the stable packages should point to the minor version I think.

File: Directory.Build.targets (hunk 1 above)
Comment (michaelgsharp) [reply Directory.Build.targets:42]: Deleted.

File: docs/samples/Microsoft.ML.Samples.GPU/Microsoft.ML.Samples.GPU.csproj (hunk 3 above)
Comment (michaelgsharp) [reply docs/samples/Microsoft.ML.Samples.GPU/Microsoft.ML.Samples.GPU.csproj:24]: To get the nuget packages to be built correctly with the correct dependencies, I had to make a lot of the project references be `PrivateAssets`. This makes them not show up as a nuget dependency. The flip side, is that any project that depends on a project with PrivateAssets has to also include the PrivateAssets projects.

File: eng/Build.props (hunk 4 above)
Comment (michaelgsharp) [reply eng/Build.props:8]: I'll run it again and see.

File: pkg/Directory.Build.props (hunk 5 above)
Comment (michaelgsharp) [reply pkg/Directory.Build.props:5]: They are obsoleted. But I wanted to leave them in until we get the symbols packages working so that we can reference anything if we need to. Is that ok? Also, the `$(ArtifactsDir)pkgassets` is where the dll files go that are to be included in each package. Its under `Artifacts/pkgassets' and is a different folder then the `pkg` folder you are referring to.

File: src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj (hunk 6 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj:3]: Yep. I was just waiting until we had the symbol packages working to remove them all. Also, I was not sure where to put that `Directory.Build.props` since we have one in most of the folders. Any thoughts?

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: I left it in because I wasn't sure if we wanted to try and find a way to fix the code I commented out or not. If you don't think it should be fixed then I can go ahead and remove all the associated code.

File: test/Directory.Build.targets (hunk 8 above)
Comment (michaelgsharp) [reply test/Directory.Build.targets:11]: Removed.

File: tools-local/Microsoft.ML.StableApi/Microsoft.ML.StableApi.csproj (hunk 9 above)
Comment (michaelgsharp) [reply tools-local/Microsoft.ML.StableApi/Microsoft.ML.StableApi.csproj:10]: Sounds good. Fixed.

File: eng/Build.props (hunk 4 above)
Comment (michaelgsharp) [reply eng/Build.props:8]: Various things fail. The native .dll hasn't finished compiling by the time its needed in the other .cs projects, so the copy fails. When the API checker runs, some of the c# .dll's aren't done yet so it fails cause the contracts is different. etc. Once its been fully built once, it works fine in parallel. But the first time it can't be. But the parallel build is just for the different main projects in the build.props file. It doesn't affect the compilation of ml.net itself. At least as far as I can tell.

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (harishsk) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: I can see a scenario where we choose to include a bunch of easy data transforms using custom transformers (like the normalizers we had discussed before). I think it is okay to remove it.

@eerhardt Any thoughts on if we this needs to be retained?


---
In reply to: [513676716](https://github.com/dotnet/machinelearning/pull/5464#discussion_r513676716) [](ancestors = 513676716)

File: eng/Build.props (hunk 4 above)
Comment (harishsk) [reply eng/Build.props:8]: Parallel builds should respect dependencies. If it is failing, it looks like we are not capturing the project dependencies correctly. Slightly lower priority work item, but we should see where it is failing. 
Can you please add either a TODO or an explicit bug for this so we don't forget this?


---
In reply to: [513684740](https://github.com/dotnet/machinelearning/pull/5464#discussion_r513684740) [](ancestors = 513684740)

Comment (michaelgsharp) [reply eng/Build.props:8]: Lets sync on this. I'm not sure that its a dependency issue.

Also, the API checker has issues even not in parallel if you run PACK with the first build. You need to run a full build without `-pack` the first time. This is probably a dependency issue, but I would like to sync on that with you briefly as well.

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (eerhardt) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: See the comment on the only callsite to this method:

https://github.com/dotnet/machinelearning/blob/d849ba4c7831586018d5b2cca6bdd2b1fa228ddf/src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs#L1077-L1079

File: src/Microsoft.Extensions.ML/Microsoft.Extensions.ML.csproj [hunk 10]
```diff
@@ -1,17 +1,27 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="$(RepoRoot)pkg/Directory.Build.props" />
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <IncludeInPackage>Microsoft.Extensions.ML</IncludeInPackage>
+    <PackageId>Microsoft.Extensions.ML</PackageId>
```
Comment (eerhardt) [inline src/Microsoft.Extensions.ML/Microsoft.Extensions.ML.csproj:6]: Why do we need to set `PackageId`? It should be defaulted to the project name.

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (harishsk) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: In Arcade, we are signing product and test binaries with the same key. Is there any other way to distinguish product binaries from test binaries? If not, do we lose anything from scanning product binaries as well (apart from a small perf hit)?




---
In reply to: [513810237](https://github.com/dotnet/machinelearning/pull/5464#discussion_r513810237) [](ancestors = 513810237)

Comment (eerhardt) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: > Is there any other way to distinguish product binaries from test binaries?

The check also checks the name of the assembly starts with `Microsoft.ML`. Another option would be to change the strong name key for the single test assembly that needs this. (It doesn't really matter which strong name we give our test assemblies)

> apart from a small perf hit

I have not measured it, so I'm not certain I would say it is "small". It affects startup. Our assemblies contain a lot of types, which we know do not contain this attribute. So it is 100% wasteful to do this scanning.

Comment (harishsk) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: @michaelgsharp Can you please try if changing the strong name key works?

---
In reply to: [513823854](https://github.com/dotnet/machinelearning/pull/5464#discussion_r513823854) [](ancestors = 513823854)

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj (hunk 2 above)
Comment (michaelgsharp) [reply docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:11]: Resolving per offline discussion.

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: Fixed the checked so it excludes Our test projects.

File: src/Microsoft.Extensions.ML/Microsoft.Extensions.ML.csproj (hunk 10 above)
Comment (michaelgsharp) [reply src/Microsoft.Extensions.ML/Microsoft.Extensions.ML.csproj:6]: You are correct. I saw this in one of the arcade examples shared with me and didn't realize it was the default behavior. I will go ahead and remove this from all packages that have it.

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (harishsk) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs]: Can you please add a note here about the change. Something along the lines that we used to have two separate strong name keys for product and test binaries and Arcade does not support that anymore (and that did not accept our PR for adding the additional signed key and using two strong name keys without the PR) ?

---
In reply to: [516842984](https://github.com/dotnet/machinelearning/pull/5464#discussion_r516842984) [](ancestors = 516842984)

**Review by harishsk — APPROVED** [review]
:shipit:

**Review by harishsk — APPROVED** [review]
:shipit:

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs (hunk 7 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1110]: Comment added.

File: src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj (hunk 6 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj:3]: Removed.

File: eng/Build.props (hunk 4 above)
Comment (michaelgsharp) [reply eng/Build.props:8]: Resolved per offline discussion.

File: src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs [hunk 11]
```diff
@@ -1110,10 +1110,15 @@ private void LoadExtensions(Assembly assembly, bool throwOnError)
...
         private static bool CanContainExtensions(Assembly assembly)
         {
+            // We used to have two separate strong name keys for product and test binaries and Arcade
+            // does not support that anymore (and that did not accept our PR for adding the additional
+            // signed key and using two strong name keys without the PR). If they ever decide to support
+            // this scenario we can remove the regex and go back to checking just the public key and startswith.
             if (assembly.FullName.StartsWith("Microsoft.ML.", StringComparison.Ordinal)
-                && HasMLNetPublicKey(assembly))
+               && !Regex.IsMatch(assembly.FullName, @"Microsoft\.ML.*\.Tests.*")
```
Comment (eerhardt) [inline src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1118]: Instead of changing product code for this (which is basically slowing down the happy path), we can just change the 1 test assembly that tests this functionality. Either change the test assembly to:

1. Not start with `Microsoft.ML.`
2. Change its public key to something other than what the product uses. This can easily be done by setting the `StrongNameKeyId` property on the test project to something other than `Open`.

Comment (michaelgsharp) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1118]: When I change the `StrongNameKeyId` I get an error because of the `InternalsVisibleTo` in the `AssemblyInfo`. I can make the change as long as I can find one of the actual values for the public keys found [here](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/StrongName.targets), but I am not sure where to find them. Do you know by chance?

Comment (eerhardt) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1118]: The way I typically find that information is by using the [`sn.exe` tool](https://docs.microsoft.com/en-us/dotnet/framework/tools/sn-exe-strong-name-tool) in the Visual Studio Developer Command Prompt. I believe the options you want are:

`sn.exe -Tp <path to .dll>`

Another option is to pull out the extensions used in the tests to a separate assembly.

Comment (harishsk) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1118]: @eerhardt The correct fix was in [this PR](https://github.com/dotnet/arcade/pull/6203) that didn't get merged to Arcade. 
Arcade does the build and signing together, so it needs the keys as part of the environment. I couldn't find any way to override it from outside Arcade (even after consultation with the Arcade folks). Checking the test signing key into Arcade is the only fix that worked and that PR not accepted. So we are now signing all binaries with the same key. Please let us know if you have a workaround.

Comment (eerhardt) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1118]: Arcade already has more than one public key that can be used. You can just pick a different one for that one test assembly.

Or another option is to put the types that have extensions in them in a separate assembly that doesn't start with `Microsoft.ML.`.

Comment (michaelgsharp) [reply src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs:1118]: [Here]( https://github.com/dotnet/machinelearning/pull/5475) is the PR that fixes the issue @eerhardt, I have requested your review on it. It is adding back in the test cert we were using before since we already had all the info for that cert.
